### PR TITLE
[Sparse] Remove SparseMatrix.indices and add device/dtype conversions

### PR DESCRIPTION
**Pull request by mufeili**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/sparse/diag_matrix.py [hunk 1]
```diff
@@ -166,6 +166,163 @@ def transpose(self):
...
         return DiagMatrix(self.val, self.shape[::-1])
 
+    def to(self, device=None, dtype=None):
+        """Perform matrix dtype and/or device conversion. If the target device
+        and/or dtype are already in use, the original matrix will be returned.
+
+        Parameters
+        ----------
+        device : torch.device, optional
+            This is the target device of the matrix. If not given, the current
```
Comment (jermainewang) [inline python/dgl/sparse/diag_matrix.py:176]: This is the target ... -> The target ...

File: python/dgl/sparse/sparse_matrix.py [hunk 2]
```diff
@@ -194,6 +171,190 @@ def transpose(self):
...
+        """
+        if device is None:
+            device = self.device
+        if dtype is None:
+            dtype = self.dtype
+
+        if device == self.device and dtype == self.dtype:
+            return self
+
+        row, col = self.coo()
```
Comment (jermainewang) [inline python/dgl/sparse/sparse_matrix.py:213]: Use val_like instead of getting coo explicitly. COO may not be materialized.

Comment (mufeili) [reply python/dgl/sparse/sparse_matrix.py:213]: This only works if the device stays unchanged.

File: python/dgl/sparse/sparse_matrix.py [hunk 3]
```diff
@@ -194,6 +172,192 @@ def transpose(self):
...
+        if device == self.device and dtype == self.dtype:
+            return self
+        elif device == self.device:
+            return val_like(self, self.val.to(dtype=dtype))
+        else:
+            row, col = self.coo()
+            row = row.to(device=device)
+            col = col.to(device=device)
+            val = self.val.to(device=device, dtype=dtype)
+            return from_coo(row, col, val, self.shape)
```
Comment (jermainewang) [inline python/dgl/sparse/sparse_matrix.py:220]: Still don't think this is the right way. Perhaps it needs a C++ operator, which will copy/cast whatever formats exist in the sparse matrix to the target.

File: python/dgl/sparse/diag_matrix.py [hunk 4]
```diff
@@ -166,6 +167,163 @@ def transpose(self):
...
+        if dtype is None:
+            dtype = self.dtype
+
+        if device == self.device and dtype == self.dtype:
+            return self
+
+        return diag(self.val.to(device=device, dtype=dtype), self.shape)
+
+    def cuda(self):
+        """Move the matrix to GPU. If the matrix is already on GPU, the
```
Comment (frozenbugs) [inline python/dgl/sparse/diag_matrix.py:208]: Will this always pick cuda:0 if multiple devices exist?

File: python/dgl/sparse/sparse_matrix.py (hunk 3 above)
Comment (frozenbugs) [reply python/dgl/sparse/sparse_matrix.py:220]: I am not sure what's the difference here. And the dtype change is (likely) not a heavy ops in the training cycle, so probably not worth to optimize more?

File: python/dgl/sparse/sparse_matrix.py [hunk 5]
```diff
@@ -194,6 +172,192 @@ def transpose(self):
...
+
+    def cuda(self):
```
Comment (frozenbugs) [inline python/dgl/sparse/sparse_matrix.py:222]: Ditto.

File: python/dgl/sparse/diag_matrix.py (hunk 4 above)
Comment (frozenbugs) [reply python/dgl/sparse/diag_matrix.py:208]: If so, add comment.

**Review by frozenbugs — APPROVED** [review]

File: python/dgl/sparse/diag_matrix.py [hunk 6]
```diff
@@ -166,6 +167,163 @@ def transpose(self):
...
+
+        >>> val = torch.ones(2)
+        >>> mat = diag(val)
+        >>> mat.float()
+        DiagMatrix(values=tensor([1., 1.]),
+                   size=(2, 2))
+        """
+        return self.to(dtype=torch.float)
+
+    def double(self):
```
Comment (frozenbugs) [inline python/dgl/sparse/diag_matrix.py:267]: not sure why do we need a list of APIs for fast dtype conversion.
The naming should be either toDouble() or asDouble() depends on whether this is a in-place conversion or not.

Comment (frozenbugs) [reply python/dgl/sparse/diag_matrix.py:267]: If this is a pytorch convention then ignore my comment.

File: python/dgl/sparse/sparse_matrix.py (hunk 3 above)
Comment (jermainewang) [reply python/dgl/sparse/sparse_matrix.py:220]: Discussed with @frozenbugs, we will table the optimization to post v1.0. cc @mufeili

**Review by jermainewang — APPROVED** [review]

File: python/dgl/sparse/sparse_matrix.py (hunk 3 above)
Comment (frozenbugs) [reply python/dgl/sparse/sparse_matrix.py:220]: Filed https://github.com/dmlc/dgl/issues/5119 to track.

**Pull request merged**